[SPARK-9468][Yarn][Core] Avoid scheduling tasks on preemption executors

jerryshao opened the pull request:
For Spark running on Yarn, if preemption is enabled by Yarn, containers will be preempted by Yarn, this will be notified to AM ahead of time. Spark itself should be aware of this preemption information and avoid scheduling tasks on these executors.

jerryshao committed:
Add preemption mechanism for Spark on Yarn

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39019/consoleFull) for   PR 7786 at commit [`6d030ec`](https://github.com/apache/spark/commit/6d030ec8a9b1e616c2c7c17697de7c5f703bd2af).

SparkQA commented:
[Test build #39019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39019/console) for   PR 7786 at commit [`6d030ec`](https://github.com/apache/spark/commit/6d030ec8a9b1e616c2c7c17697de7c5f703bd2af).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class PreemptionExecutors(executorIds: Set[String]) extends CoarseGrainedClusterMessage`
  * `        The class name of the paired Scala UDT (could be '', if there`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39178 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39178/consoleFull) for   PR 7786 at commit [`90dd2f8`](https://github.com/apache/spark/commit/90dd2f8aa80ed3b52047e194c053976a5bbe0fad).

SparkQA commented:
[Test build #39178 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39178/console) for   PR 7786 at commit [`90dd2f8`](https://github.com/apache/spark/commit/90dd2f8aa80ed3b52047e194c053976a5bbe0fad).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -94,4 +94,5 @@ private[spark] object CoarseGrainedClusterMessages {
 
   case class KillExecutors(executorIds: Seq[String]) extends CoarseGrainedClusterMessage
 
+  case class PreemptionExecutors(executorIds: Set[String]) extends CoarseGrainedClusterMessage
```
`PreemptExecutors` to be consistent with the other messages?

vanzin replied:
+1. In fact, since you're sending a message saying which executors are about to be pre-empted, a better name would be `PreemptedExecutors`. Same thing everywhere else in this patch.

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala:
```diff
@@ -111,6 +124,16 @@ private[spark] abstract class YarnSchedulerBackend(
+
+        synchronized {
+          preemptionExecutorIDs.clear()
+          preemptionExecutorIDs ++= executorIds
```
`preemptedExecutorIDs`

CodingCat commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -338,6 +338,7 @@ private[spark] class ApplicationMaster(
     val t = new Thread {
       override def run() {
         var failureCount = 0
+        var lastPreemptionExecutors: Set[String] = Set.empty
```
`lastPreemptedExecutors`?

CodingCat commented:
is the same problem existing in Mesos?

jerryshao commented:
Thanks a lot @CodingCat for your review, I will keep consistent about variable names. Besides I'm not familiar with Mesos, if Mesos supports resource preemption, I think this problem also exists in Mesos.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #42340 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42340/consoleFull) for   PR 7786 at commit [`90dd2f8`](https://github.com/apache/spark/commit/90dd2f8aa80ed3b52047e194c053976a5bbe0fad).

SparkQA commented:
[Test build #42340 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42340/console) for   PR 7786 at commit [`90dd2f8`](https://github.com/apache/spark/commit/90dd2f8aa80ed3b52047e194c053976a5bbe0fad).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `  case class PreemptionExecutors(executorIds: Set[String]) extends CoarseGrainedClusterMessage`

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42340/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -191,7 +194,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     private def makeOffers(executorId: String) {
       // Filter out executors under killing
-      if (!executorsPendingToRemove.contains(executorId)) {
+      if (!executorsPendingToRemove.contains(executorId) && !preemptionExecutors(executorId)) {
```
Same result, but I think explicitly calling `.contains(executorId)` is clearer.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -346,7 +347,16 @@ private[spark] class ApplicationMaster(
+              // list.
+              val preemptionExecutors = allocator.getPreemptionExecutors
+              if (preemptionExecutors != lastPreemptionExecutors) {
+                amEndpoint.send(PreemptionExecutors(preemptionExecutors))
```
If you plumb `driverRef` from `registerAM` into this method (`launchReporterThread`), then you can send the message to the driver directly, instead of making an extra hop in the AM for no reason.

jerryshao replied:
I would incline to use `amEndpoint` here as to keep consistent with other places like `AddWebFIlter`.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -238,6 +257,33 @@ private[yarn] class YarnAllocator(
+
+    // We need to forget all the already preempted executors at this time,
+    // because preemption information will be changed time to time.
+    preemptionExecutors.clear()
```
This is not thread-safe. You can be going through the list in the `ApplicationMaster` when this code runs, and that might cause exceptions.

Instead, I'd recommend returning the set of preempted executors from `updateResources`, always creating a new set when this method is called. Then you also don't need the class field nor the synchronized getter.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -238,6 +257,33 @@ private[yarn] class YarnAllocator(
 
+  /** Get a list of preemption containers through AM-RM heartbeat. */
+  private def preemptResources(allocateResponse: AllocateResponse): Unit = {
+    val preemptMessageOpt = Option(allocateResponse.getPreemptionMessage)
```
Part of me things these `Option` instances are not necessary and good old null checks would be just as readable, but I don't really mind either way.

vanzin commented:
@jerryshao could you rebase the PR and address the feedback?

One thing that would be nice is to actually take action on the driver to get rid of executors that are in the "strict" set if they are idle. That can help YARN in freeing resources earlier, and could also allow for a replacement executor on a non-contended node to be started earlier, since the preemption cycle can take some time. These executors could be in this "preempted" state for a long time (several seconds), which is enough to run a lot of small tasks.

jerryshao commented:
Thanks @vanzin for your review, I will update the code according to your comments.

jerryshao commented:
Yes, releasing some idle executors to alleviate the preemption in RM is a good advise, basically I think there're two aspect of things we can handle about preemption:

1. Avoid scheduling tasks on to-be preempted executors, this is what I addressed in this PR.
2. Release some idle nodes to alleviate preemption.

I think this existing implementation might be too simple to handle the problem gracefully, I will keep improving it.

jerryshao committed:
Address the comments

jerryshao committed:
Add unit test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

jerryshao committed:
Style fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42940/consoleFull) for   PR 7786 at commit [`41bebec`](https://github.com/apache/spark/commit/41bebec6067ccdcb1b1acb31894fff4c6249ceb2).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42939/
Test FAILed.

jerryshao commented:
Hi @vanzin , I just updated the code and added unit test. Would you please help to review again? Thanks a lot.

Also I discussed with Yarn guys about how to use `PreemptionResourceRequests` to give Spark a hint to release some idle executors in advance, looks like currently Yarn not well supports this feature, and the behavior is different between schedulers, so I think it would be better to remain as a open question to think about. Besides introducing this mechanism still requires more design and implementation, I'd like to address this in a separate PR if there's a good design thought.

SparkQA commented:
[Test build #42940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42940/console) for   PR 7786 at commit [`41bebec`](https://github.com/apache/spark/commit/41bebec6067ccdcb1b1acb31894fff4c6249ceb2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class PreemptExecutors(executorId: Set[String]) extends CoarseGrainedClusterMessage`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42940/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -449,6 +458,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+   * @return A set of executor IDs
+   */
+  protected def preemptedExecutors: Set[String] = synchronized {
+    preemptedExecutorIDs.toSet
```
Why not store it as a `Set` in the first place? It's not like you're modifying it.

vanzin replied:
I see you modify it in the test, but I think that's the wrong thing to do. Better to avoid this transformation or extra allocation or whatever overhead, since this method is called pretty often by the scheduling code.

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,113 @@
+
+    val backend = sc.schedulerBackend
+    val executors = {
+      val method = backend.getClass.getSuperclass.getDeclaredMethod(
```
Hmmm, no. If you need access to this field, make it visible, or use `PrivateMethodTester`.

jerryshao replied:
I'm not sure if `PrivateMethodTester` can be worked for superclass, I did tried before, but seems have some problems to find out the method. Seems if you don't accept reflection way, make it visible is the only solution.

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,113 @@
+    assert(executors.size === 2)
+
+    val preemptedExecutors = {
+      val method = backend.getClass.getSuperclass.getDeclaredMethod("preemptedExecutorIDs")
```
Don't use reflection here. Just make fields visible.

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,113 @@
+}
+
+private object CoarseGrainedSchedulerBackendSuite {
+  class TaskStateTracker extends SparkListener {
```
Why not just declare this inside the class? No need for the `object`.

jerryshao committed:
Address the comments

jerryshao commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43091 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43091/consoleFull) for   PR 7786 at commit [`57427f7`](https://github.com/apache/spark/commit/57427f731a2d69c6536c1e7cd1d7af185f64e348).

SparkQA commented:
[Test build #43091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43091/console) for   PR 7786 at commit [`57427f7`](https://github.com/apache/spark/commit/57427f731a2d69c6536c1e7cd1d7af185f64e348).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43091/
Test PASSed.

mccheah commented:
Quick question - Is it guaranteed that if the AM is told "Container X may be preempted" it necessitates that "Container X will definitely be preempted"? We should link documentation to this behavior somewhere. If not, how does our code handle if the YARN scheduler changes its mind about which containers to preempt?

jerryshao commented:
Hi @mccheah sorry for late response, looks like this message is out of my radar.

From Yarn's side, Yarn makes this preemption decision depends on status of that time in RM, the status will be changed (other applications may free the container), so the preempted containers will be changed time to time, no guarantee that this container will definitely be preempted.

Here in my solution, I also updates the preempted containers time to time, so if at this time, container A is preempted, tasks will be avoided to allocate to container A. But later on when A is not preempted, tasks can be allocated on container A again.

The preemption information only reflects the status of cluster at that time, so in the driver side this preemption information will be updated time to time to reflect this changing.

steveloughran commented:
Jerry: do you know when YARN visits this decisions about having things pre-emptible? That is: if you are given a warning dos that mean the container will go any time in the next 60s? hour? As the longer the window is, the less useful it is avoiding scheduling work —you lose too much.

Or is this really a warning that YARN wants the container back and that you should just release it gracefully? If that's the case, then graceful shutdown of work is exactly what's needed

jerryshao commented:
@steveloughran , by default Yarn will preempt the container 15 seconds after the warning. Yarn RM gives the warning to AM depends on the cluster usage at that time, it will be changed time to time, once the preemption list is changed, this executor could be rescheduled again.

Also it could not be said that this preempted container will definitely be reclaimed by Yarn, Yarn will change the decision time to time, so using the old decision to release the containers may not be a good idea.

vanzin commented:
> by default Yarn will preempt the container 15 seconds after the warning

That's a long time and you can run a lot of tasks in that time. Unless Spark actively goes and gets rid of these executors to play nice, this change would probably harm performance more than it helps. So maybe instead we should focus on stopping the executors so that dynamic allocation can bring up new ones elsewhere?

Otherwise the current state is probably better.

steveloughran commented:
@vanzin I suspect that if you get told you are being pre-empted, you aren't likely to get containers elsewhere —pre-emption is a sign of demand being too high, and your queue lower priority. But pre-requesting a new container while continuing the current work might be a nice trick, keeping the live executors busy while queueing up early requests for the replacements

vanzin commented:
@steveloughran 

> I suspect that if you get told you are being pre-empted, you aren't likely to get containers elsewhere

That's very possible. But I'm just trying to point out that the current change doesn't really make things better. Without killing the executor, you'll still be holding on to resources, except now you wouldn't be using them. So might as well keep using them for as long as you can, or give them back as soon as possible (i.e. as soon as the executor becomes idle).

steveloughran commented:
> But I'm just trying to point out that the current change doesn't really make things better. Without killing the executor, you'll still be holding on to resources, except now you wouldn't be using them. So might as well keep using them for as long as you can, or give them back as soon as possible (i.e. as soon as the executor becomes idle).

I agree. Doing a clean shutdown and release is probably what the schedulers prefer —especially if it starts up the pre-empting code faster

vanzin commented:
@jerryshao do you plan to implement either of the above suggestions? otherwise we should probably close this PR.

jerryshao commented:
@vazin, sorry for the delay. 

I think your concern is valid that this change will probably harm the performance since 15 seconds (preemption waiting) is not a short time. 

But your suggestion of killing the executors ahead of preemption to release the resource is also risky, in Yarn's side preemption is based on the decision at that time, the decision will be changed time to time since the resource is dynamically changed, if later on the resource is enough that don't need to do preemption, this pre-killing and re-acquire of containers will also harm the performance to some extent.

Maybe current implementation is sufficient enough to address the preemption problem, we don't need to do anything about it. I'm going to close this, thanks a lot for your time and review.

chemikadze commented:
@vanzin If those would be implemented, would it have any change to get merged? We use preemption quite a lot and current behavior is not the best we can get: logs sometimes getting overfilled with preemption side effects (RPC errors, etc), getting logs hard to read and confusing some users. I agree that depending on task size, effect might be both positive and negative (longer ones anyway won't be able to complete be wasting resources, but lots of shorter ones will not get chance to run). Does it just mean it should be configurable behavior (spark.yarn.releasePreemptedContainers=true/false)?
